File: src/parser/svg/ClosePath.java
package parser.svg;

import java.util.List;
import shapes.Line;
import shapes.Shape;

class ClosePath {

  // Parses close path commands (Z and z commands)
  private static List<Shape> parseClosePath(SvgState state, List<Float> args) {
    if (!state.currPoint.equals(state.initialPoint)) {
      Line line = new Line(state.currPoint, state.initialPoint);
      state.currPoint = state.initialPoint;
      return List.of(line);
    } else {
      return List.of();
    }
  }

  static List<Shape> absolute(SvgState state, List<Float> args) {
    return parseClosePath(state, args);
  }

  static List<Shape> relative(SvgState state, List<Float> args) {
    return parseClosePath(state, args);
  }
}

File: Source/svg/ClosePath.cpp
#include "ClosePath.h"
#include "../shape/Line.h"

std::vector<std::unique_ptr<Shape>> ClosePath::absolute(SvgState& state, std::vector<float>& args) {
	return parseClosePath(state, args);
}

std::vector<std::unique_ptr<Shape>> ClosePath::relative(SvgState& state, std::vector<float>& args) {
	return parseClosePath(state, args);
}

std::vector<std::unique_ptr<Shape>> ClosePath::parseClosePath(SvgState& state, std::vector<float>& args) {
	auto shapes = std::vector<std::unique_ptr<Shape>>();
	if (state.currPoint.x != state.initialPoint.x || state.currPoint.y != state.initialPoint.y) {
        state.currPoint = state.initialPoint;
		shapes.push_back(std::make_unique<Line>(state.currPoint.x, state.currPoint.y, state.initialPoint.x, state.initialPoint.y));
    }
	return shapes;
}